pkg/userdata/generate/generate.go

```diff
@@ -0,0 +1,223 @@
+
+	b := make([]rune, n)
+	for i := range b {
+		b[i] = letter[rand.Intn(len(letter))]
+	}
+	return string(b)
+}
+
+// NewInput generates the sensitive data required to generate all userdata
+// types.
+// nolint: gocyclo
+func NewInput(clustername string, masterIPs []string) (input *Input, err error) {
+	kubeadmTokens := &KubeadmTokens{
+		BootstrapToken: RandomString(6) + "." + RandomString(16),
+		CertKey:        RandomString(26),
+	}
+
+	trustdInfo := &TrustdInfo{
+		Username: RandomString(14),
+		Password: RandomString(24),
+	}
+
+	// Generate Kubernetes CA.
+	opts := []x509.Option{x509.RSA(true), x509.Organization("talos-k8s")}
+	k8sCert, err := x509.NewSelfSignedCertificateAuthority(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	// Generate Talos CA.
+	opts = []x509.Option{x509.RSA(false), x509.Organization("talos-os")}
+	osCert, err := x509.NewSelfSignedCertificateAuthority(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	// Generate the admin talosconfig.
+	adminKey, err := x509.NewKey()
+	if err != nil {
+		return nil, err
+	}
+	pemBlock, _ := pem.Decode(adminKey.KeyPEM)
+	if pemBlock == nil {
+		return nil, errors.New("failed to decode admin key pem")
+	}
+	adminKeyEC, err := stdlibx509.ParseECPrivateKey(pemBlock.Bytes)
+	if err != nil {
+		return nil, err
+	}
+	ips := []net.IP{net.ParseIP("127.0.0.1")}
+	opts = []x509.Option{x509.IPAddresses(ips)}
+	csr, err := x509.NewCertificateSigningRequest(adminKeyEC, opts...)
+	if err != nil {
+		return nil, err
+	}
+	csrPemBlock, _ := pem.Decode(csr.X509CertificateRequestPEM)
+	if csrPemBlock == nil {
+		return nil, errors.New("failed to decode csr pem")
+	}
+	ccsr, err := stdlibx509.ParseCertificateRequest(csrPemBlock.Bytes)
+	if err != nil {
+		return nil, err
+	}
+	caPemBlock, _ := pem.Decode(osCert.CrtPEM)
+	if caPemBlock == nil {
+		return nil, errors.New("failed to decode ca cert pem")
+	}
+	caCrt, err := stdlibx509.ParseCertificate(caPemBlock.Bytes)
+	if err != nil {
+		return nil, err
+	}
+	caKeyPemBlock, _ := pem.Decode(osCert.KeyPEM)
+	if caKeyPemBlock == nil {
+		return nil, errors.New("failed to decode ca key pem")
+	}
+	caKey, err := stdlibx509.ParseECPrivateKey(caKeyPemBlock.Bytes)
+	if err != nil {
+		return nil, err
+	}
+	adminCrt, err := x509.NewCertificateFromCSR(caCrt, caKey, ccsr)
+	if err != nil {
+		return nil, err
+	}
+
+	certs := &Certs{
+		AdminCert: base64.StdEncoding.EncodeToString(adminCrt.X509CertificatePEM),
+		AdminKey:  base64.StdEncoding.EncodeToString(adminKey.KeyPEM),
+		OsCert:    base64.StdEncoding.EncodeToString(osCert.CrtPEM),
+		OsKey:     base64.StdEncoding.EncodeToString(osCert.KeyPEM),
+		K8sCert:   base64.StdEncoding.EncodeToString(k8sCert.CrtPEM),
+		K8sKey:    base64.StdEncoding.EncodeToString(k8sCert.KeyPEM),
+	}
+
+	input = &Input{
+		Certs:         certs,
+		MasterIPs:     masterIPs,
+		PodNet:        []string{"10.244.0.0/16"},
+		ServiceNet:    []string{"10.96.0.0/12"},
+		ServiceDomain: "cluster.local",
```

Comment:
I'm fine with leaving these in as a first pass, but I'm expecting these bits would be some of the more common changes.

Reply:
Agree. This PR is mostly just moving what we have in the Cluster API provider over to hear. I plan on expanding it in a follow-up PR.